src/mlpack/methods/grad_boosting/grad_boosting.hpp

```diff
@@ -0,0 +1,217 @@
+#ifndef MLPACK_METHODS_GRAD_BOOSTING_GRADBOOSTING_HPP
+#define MLPACK_METHODS_GRAD_BOOSTING_GRADBOOSTING_HPP
+
+// Importing base components required to write mlpack methods.
+#include <mlpack/core.hpp>
+
+// Only using decision trees for now, therefore only including
+// decision tree functionalities.
+#include <mlpack/methods/decision_tree/decision_tree_regressor.hpp>
+
+
+// Written in mlpack namespace.
+namespace mlpack {
+/**
+ * The Gradient Boosting class. Gradient Boosting is a boosting algorithm, meaning that it
+ * combines an ensemble of weak learners to produce a strong learner.
+ * 
+ * Gradient Boosting is generally implemented using Decision Trees, or more specifically 
+ * Decision Stumps i.e. weak learner Decision Trees with low depth.
+ * 
+ * @tparam MatType Data matrix type (i.e. arma::mat or arma::sp_mat).
+ */
+
+
+template<typename MatType = arma::mat>
+class GradBoosting
+{
+ public:
+  // Convenience typedef for the type of weak learner we are using.
+  typedef mlpack::DecisionTreeRegressor<mlpack::MSEGain,
+                      mlpack::BestBinaryNumericSplit,
+                      mlpack::AllCategoricalSplit,
+                      mlpack::AllDimensionSelect,
+                      false> WeakLearnerType;
+
+  typedef typename MatType::elem_type ElemType;
+
+  /**
+   * Constructor for creating GradBoosting without training. 
+   * Be sure to call Train() before calling Classify()
+   */
+  GradBoosting();
+
+  /**
+   * Constructor in case weak learners aren't defined.
+   *
+   * @param data Input data
+   * @param labels Corresponding labels
+   * @param numClasses Number of classes
+   * @param numWeakLearners Number of weak learners
+   */
+  GradBoosting(const MatType& data,
+               const data::DatasetInfo& datasetInfo,
+               const arma::Row<size_t>& labels,
+               const size_t numClasses,
+               const size_t numWeakLearners);
+
+  /**
+   * Constructor for a GradBoosting model. Any extra parameters are used as
+   * hyperparameters for the weak learner. These should be the last arguments
+   * to the weak learner's constructor or `Train()` function (i.e. anything
+   * after `numClasses` or `weights`).
+   *
+   * @param data Input data.
+   * @param labels Corresponding labels.
+   * @param numClasses The number of classes.
+   * @param numWeakLearners Number of weak learners.
+   * @param minimumLeafSize Minimum number of points in each leaf node.
+   * @param minimumGainSplit Minimum gain for the node to split.
+   * @param maximumDepth Maximum depth for the tree.
+   */
+  GradBoosting(const MatType& data,
+               const data::DatasetInfo& datasetInfo,
+               const arma::Row<size_t>& labels,
+               const size_t numClasses,
+               const size_t numWeakLearners,
+               const size_t minimumLeafSize,
+               const double minimumGainSplit,
+               const size_t maximumDepth);
```

Comment:
I think we need default arguments for all of these methods?  If so, then the constructor above (with only `numClasses` and `numWeakLearners`) is not needed.  Other classes have default arguments for anything starting after `numClasses`.

Also, the overloads you have for the constructor and training only accept mixed categorical data (i.e. they have a `datasetInfo` parameter).  For more information about how we represent numerical vs. categorical data in mlpack see [this documentation page](https://www.mlpack.org/doc/user/load_save.html).  So we should also have overloads that *don't* take a `datasetInfo` parameter (but are otherwise the same), which then assume that the data is all numeric in type.